File: Desafios Aleatorios/Python/exe04.py
USUARIO_REGISTRADO = "usuario"
SENHA_REGISTRADA = "senha"
TENTATIVAS = 3

def titulo(texto: str):
    print("\n" + "=" * 40)
    print(texto.center(40))
    print("=" * 40)

def obter_usuario() -> str:
    return input("\nDigite seu nome: ")

def validar_usuario(nome: str) -> bool:
    return nome == USUARIO_REGISTRADO

def obter_senha() -> str:
    return input("\nDigite sua senha: ")

def validar_senha(senha: str) -> bool:
    return senha == SENHA_REGISTRADA

def liberar_acesso() -> bool:
    tentativas = TENTATIVAS
    titulo("LOGIN - CAIXA ELETRONICO")
    
    while tentativas:
        print(f"\nTentativas restantes: {tentativas}")
        usuario = obter_usuario()
        
        if not validar_usuario(usuario):
            tentativas -= 1
            print("Usuario invalido.")
        else:
            while tentativas:
                print(f"\nTentativas restantes: {tentativas}")
                senha = obter_senha()
                if not validar_senha(senha):
                    tentativas -= 1
                    print("Senha invalida.")
                else:
                    print("\nAcesso liberado com sucesso!\n")
                    return True
            print("\nAcesso negado. Encerrando...\n")
            return False
    
    print("\nAcesso negado. Encerrando...\n")
    return False

def exibir_menu() -> str:
    titulo("MENU PRINCIPAL")
    print("1. Sacar")
    print("2. Depositar")
    print("3. Exibir Saldo")
    print("0. Sair")
    return input("\nEscolha uma opcao (0-3): ")

def validar_escolha(escolha: str) -> bool:
    try:
        escolha = int(escolha)
        if escolha < 0 or escolha > 3:
            print("Opcao fora do intervalo (0 a 3).")
            return False
        return True
    except ValueError:
        print("Entrada invalida. Digite um numero.")
        return False

def validar_positivos(n: float|int) -> bool:
    return n > 0

if __name__ == "__main__":
    if not liberar_acesso():
        exit()
    
    try:
        saldo = float(input("\nInforme o saldo inicial da conta: "))
        if not validar_positivos(saldo):
            print("Informe apenas valores positivos.")
            exit()
    except Exception as erro:
        print(f"Erro: {str(erro)}")
        exit()

    while True:
        escolha = exibir_menu()
        if not validar_escolha(escolha):
            print("Tente novamente.")
            continue

        escolha = int(escolha)

        match escolha:
            case 0:
                print("\nObrigado, volte sempre!\n")
                break
            case 1:
                try:
                    saque = float(input("\nInforme o valor de saque: "))
                    if not validar_positivos(saque):
                        print("Valor de saque deve ser positivo.")
                        continue
                    if saldo < saque:
                        print("Saldo insuficiente para saque.")
                        continue
                    saldo -= saque
                    print(f"Saque realizado. Saldo atual: R$ {saldo:,.2f}")
                except ValueError:
                    print("Entrada invalida.")
            case 2:
                try:
                    deposito = float(input("\nInforme o valor de deposito: "))
                    if not validar_positivos(deposito):
                        print("Valor de deposito deve ser positivo.")
                        continue
                    saldo += deposito
                    print(f"Deposito realizado. Saldo atual: R$ {saldo:,.2f}")
                except ValueError:
                    print("Entrada invalida.")
            case 3:
                print(f"\nSaldo Atual: R$ {saldo:,.2f}\n")

File: Desafios Aleatorios/C/Exe04.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#define USUARIO_REGISTRADO "usuario"
#define SENHA_REGISTRADA "senha"
#define TENTATIVAS 3

void titulo(const char *texto) {
    printf("\n========================================\n");
    printf("%s\n", texto);
    printf("========================================\n");
}

void obterUsuario(char usuario[], int tamanho) {
    printf("\nDigite seu nome: ");
    scanf("%s", usuario);
}

int validarUsuario(const char usuario[]) {
    return strcmp(usuario, USUARIO_REGISTRADO) == 0;
}

void obterSenha(char senha[], int tamanho) {
    printf("\nDigite sua senha: ");
    scanf("%s", senha);
}

int validarSenha(const char senha[]) {
    return strcmp(senha, SENHA_REGISTRADA) == 0;
}

int liberarAcesso() {
    int tentativas = TENTATIVAS;
    char usuario[50], senha[50];

    titulo("LOGIN - CAIXA ELETRONICO");

    while (tentativas > 0) {
        printf("\nTentativas restantes: %d\n", tentativas);
        obterUsuario(usuario, sizeof(usuario));
        if (!validarUsuario(usuario)) {
            tentativas--;
            printf("Usuario invalido.\n");
        } else {
            while (tentativas > 0) {
                printf("\nTentativas restantes: %d\n", tentativas);
                obterSenha(senha, sizeof(senha));
                if (!validarSenha(senha)) {
                    tentativas--;
                    printf("Senha invalida.\n");
                } else {
                    printf("\nAcesso liberado com sucesso!\n\n");
                    return 1;
                }
            }
            printf("\nAcesso negado. Encerrando...\n");
            return 0;
        }
    }
    printf("\nAcesso negado. Encerrando...\n");
    return 0;
}

int validarPositivo(float n) {
    return n > 0;
}

int main() {
    if (!liberarAcesso()) {
        exit(1);
    }

    float saldo;
    printf("\nInforme o saldo inicial da conta: ");
    if (scanf("%f", &saldo) != 1 || !validarPositivo(saldo)) {
        printf("Informe apenas valores positivos.\n");
        exit(1);
    }

    while (1) {
        titulo("MENU PRINCIPAL");
        printf("1. Sacar\n2. Depositar\n3. Exibir Saldo\n0. Sair\n");
        printf("\nEscolha uma opcao (0-3): ");
        int escolha;
        if (scanf("%d", &escolha) != 1) {
            printf("Entrada invalida.\n");
            while(getchar() != '\n'); // limpa buffer
            continue;
        }

        switch (escolha) {
            case 0:
                printf("\nObrigado, volte sempre!\n");
                return 0;
            case 1: {
                float saque;
                printf("\nInforme o valor de saque: ");
                if (scanf("%f", &saque) != 1 || !validarPositivo(saque)) {
                    printf("Valor de saque invalido.\n");
                    while(getchar() != '\n');
                    continue;
                }
                if (saldo < saque) {
                    printf("Saldo insuficiente para saque.\n");
                    continue;
                }
                saldo -= saque;
                printf("Saque realizado. Saldo atual: R$ %.2f\n", saldo);
                break;
            }
            case 2: {
                float deposito;
                printf("\nInforme o valor de deposito: ");
                if (scanf("%f", &deposito) != 1 || !validarPositivo(deposito)) {
                    printf("Valor de deposito invalido.\n");
                    while(getchar() != '\n');
                    continue;
                }
                saldo += deposito;
                printf("Deposito realizado. Saldo atual: R$ %.2f\n", saldo);
                break;
            }
            case 3:
                printf("\nSaldo Atual: R$ %.2f\n", saldo);
                break;
            default:
                printf("Opcao fora do intervalo (0 a 3).\n");
        }
    }

    return 0;
}
